### PR TITLE
[Notifier] Check for MercureBundle in MercureTransportFactory

### PR DESCRIPTION
**Pull request by mtarld**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        |

https://github.com/symfony/symfony/pull/39903 has removed the check for `MercureBundle` from the `FrameworkExtension`.
The following PR is re-adding that check but in the `MercureTransportFactory` class.

### CODE REVIEW

**Review by OskarStark — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by xabbuh** [comment]
Would it make sense to add the MecureBundle as a dependency of the bridge then?

**Comment by mtarld** [comment]
Actually, if that check lives in the bridge, then that bridge won't be able to exist without the bundle. 
Therefore, the Notifier component won't be able to exist as a single component when using the Mercure bridge.
That was why I put before the check in the FrameworkExtension.

So I see two options:
- Require the bundle in composer.json and then assume that the Mercure bridge can't be used with Notifier as a single component.
- Update changes made in https://github.com/symfony/symfony/pull/39903 in order to re-do the check in FrameworkExtension (and still remove the definition if not needed)

WDYT?

File: src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php [hunk 1]
```diff
@@ -45,6 +46,10 @@ public function create(Dsn $dsn): TransportInterface
             throw new UnsupportedSchemeException($dsn, 'mercure', $this->getSupportedSchemes());
         }
 
+        if (!class_exists(MercureBundle::class)) {
+            throw new LogicException('The MercureBundle is not registered in your application. Try running "composer require symfony/mercure-bundle".');
+        }
+
         $publisherId = $dsn->getHost();
         if (!$this->publisherLocator->has($publisherId)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php:50]: Maybe we should add the check here:

```php
if (!class_exists(MercureBundle::class && 0 === \count($this->publisherLocator->getProvidedServices()))) {
    throw new LogicException(sprintf('"%s" not found. Did you forget to install the MercureBundle? Try running "composer require symfony/mercure-bundle".', $publisherId));
} else {
    throw new LogicException(sprintf('"%s" not found. Did you mean one of: %s?', $publisherId, implode(', ', array_keys($this->publisherLocator->getProvidedServices()))));
}
```

What do you think?

Comment (mtarld) [reply src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php:50]: Sounds good!

Comment (mtarld) [reply src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php:54]: Or we can check as well the presence of `FrameworkExtension`? Somthing like:
```php
        if (!$this->publisherLocator->has($publisherId)) {
            if (!class_exists(MercureBundle::class) && class_exists(FrameworkExtension::class)) {
                throw new LogicException('The MercureBundle is not registered in your application. Try running "composer require symfony/mercure-bundle".');
            }
            
            if (0 === \count($this->publisherLocator->getProvidedServices())) {
                throw new LogicException('No publisher were found. $publisherLocator is empty.');
            }

            throw new LogicException(sprintf('"%s" not found. Did you mean one of: %s?', $publisherId, implode(', ', array_keys($this->publisherLocator->getProvidedServices()))));
        }

```

What do you think?

File: src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php [hunk 2]
```diff
@@ -47,6 +48,10 @@ public function create(Dsn $dsn): TransportInterface
 
         $publisherId = $dsn->getHost();
         if (!$this->publisherLocator->has($publisherId)) {
+            if (!class_exists(MercureBundle::class) && 0 === \count($this->publisherLocator->getProvidedServices())) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php]: no need for `count()`: `!$this->publisherLocator->getProvidedServices()`

File: src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php [hunk 3]
```diff
@@ -47,6 +48,10 @@ public function create(Dsn $dsn): TransportInterface
 
         $publisherId = $dsn->getHost();
         if (!$this->publisherLocator->has($publisherId)) {
+            if (!class_exists(MercureBundle::class) && !$this->publisherLocator->getProvidedServices()) {
+                throw new LogicException('The MercureBundle is not registered in your application. Try running "composer require symfony/mercure-bundle".');
```
Comment (xabbuh) [inline src/Symfony/Component/Notifier/Bridge/Mercure/MercureTransportFactory.php]: I would reword this to something like this:

> No publishers found. Did you forget to install the MercureBundle? Try running "composer require symfony/mercure-bundle".

**Review by OskarStark — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
Thank you @mtarld.
